gh-79009: sqlite3.iterdump now correctly handles tables with autoincrement

itssme opened the pull request:
The iterdump command in Lib/sqlite3/dump.py assumed that the table "sqlite_sequence" already exists. When this was not the case and exception was thrown.

Changes:
* Sqlite3 iterdump now deletes and recreates the "sqlite_sequence" table at the end of the transaction (Just like the sqlite3 command ".dump" does)
* Added two new test for the iterdump command

Resolves gh-79009

itssme committed:
bpo-34828 sqlite3.iterdump now correctly handles tables with autoincrement

The iterdump command in Lib/sqlite3/dump.py assumed that the table "sqlite_sequence" already exists. When this was not the case and exception was thrown.

itssme committed:
bpo-34828: added entry in NEWS.d/next/Library and in ACKS

itssme committed:
Merge remote-tracking branch 'origin/master' into fix-issue-34828

# Conflicts:
#	Lib/sqlite3/test/dump.py

itssme committed:
bpo-34828 Removed redundant code and made some small formatting changes according to PEP 8

erlend-aasland commented on Misc/NEWS.d/next/Library/2018-09-28-22-18-03.bpo-34828.5Zyi_S.rst:
```diff
@@ -0,0 +1,3 @@
+Sqlite3.iterdump now correctly deletes the "sqlite_sequence" table and
+inserts the rows from the dumped db. Just like the command ".dump" in
+sqlite3 does.
```
I'd shorten this to something more accessible. For example:

    :meth:`~sqlite3.Connection.iterdump` now handles databases that use ``AUTOINCREMENT`` in one or more tables.

See the [dev guide](https://devguide.python.org/documenting/#additional-markup-constructs) for how to add reST markups.

erlend-aasland replied:
You can also add _"Patch by Joel Klimont"_ at the end of the NEWS entry.

erlend-aasland commented on Lib/sqlite3/test/dump.py:
```diff
@@ -1,7 +1,8 @@
-import unittest
 import sqlite3 as sqlite
+import unittest
+
```
Please remove any unrelated style changes from the PR. Ditto for added newlines (line 135, 144 and 149).

erlend-aasland commented on Lib/sqlite3/test/dump.py:
```diff
@@ -49,6 +50,67 @@ def test_table_dump(self):
+                        ["""INSERT INTO "sqlite_sequence" VALUES('tags',1);"""] + \
+                        ['COMMIT;']
+        [self.assertEqual(expected_sqls[i], actual_sqls[i])
+         for i in range(len(expected_sqls))]
```
Please format the new code according to [PEP 8](https://www.python.org/dev/peps/pep-0008/). Even if large parts of these files _do not_ follow PEP 8, we strive to follow PEP 8 (and PEP 7) when adding new code, or modifying existing code :)

itssme committed:
Merge branch 'main' into fix-issue-34828

# Conflicts:
#	Lib/sqlite3/test/test_dump.py

itssme committed:
bpo-34828 formatting changes as well as readability improvements in sqlite3 dump tests

erlend-aasland commented on Lib/sqlite3/dump.py:
```diff
@@ -67,4 +72,8 @@ def _iterdump(connection):
     for name, type, sql in schema_res.fetchall():
         yield('{0};'.format(sql))
 
+    # Yield statements concerning the sqlite_sequence table at the end of the transaction: (bpo-34828)
```
Is there a reason to yield these at the very end? Why not loop and yield after line 33?

itssme replied:
Currently the problem is that we assume, that the sqlite sequence table already exists. However, if we try to `delete from` the table to clear it and (by chance) it does not exist yet (no table with an autoincrement exists in the db) we get an error when running the sql commands the iterdump provides us with.

Example:
We have run the following commands on the database:
```sql
CREATE TABLE posts (id int primary key);
INSERT INTO posts VALUES(0);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(NULL,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
```

Now on the sqlite3 console we run the `.dump` command we are tying to mimic in pythons iterdump:
```sql
sqlite> .dump
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE posts (id int primary key);
INSERT INTO posts VALUES(0);
CREATE TABLE IF NOT EXISTS "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO tags VALUES(1,'test',0);
CREATE TABLE IF NOT EXISTS "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
DELETE FROM sqlite_sequence;
INSERT INTO sqlite_sequence VALUES('tags',1);
COMMIT;
```

As we can see the `DELETE FROM sqlite_sequence;` and the inserts into that table, are at the very end. If we run these sql commands on a new database, everything works as expected and the we get no error about deleting data from the sqlite sequence table, as it already exists, because the tables with the autoincrement are already created in the database.

If we now try to do the same thing in python, by creating a sql connection to a :memory: db and then creating the tables, inserting the data etc..:
```python
import sqlite3
conn = sqlite3.connect(":memory:")
cur = conn.cursor()
expected_sqls = [
             "CREATE TABLE \"posts\" (id int primary key);",
             "INSERT INTO \"posts\" VALUES(0);",
             "CREATE TABLE \"tags\" (id integer primary key autoincrement, tag varchar(256), post int references posts);",
             "INSERT INTO \"tags\" VALUES(NULL,'test',0);",
             "CREATE TABLE \"tags2\" (id integer primary key autoincrement, tag varchar(256), post int references posts);"
         ]
for sql_statement in expected_sqls:
     cur.execute(sql_statement)
iterdump = conn.iterdump()
actual_sqls = [sql_statement for sql_statement in iterdump]
for sql_command in actual_sqls:
    print(sql_command)
```

We see that the dump command already wants to delete the data from the sqlite_sequence table after creating the posts table. However, in the posts table no autoincrement is used, and the sqlite_sequence table does not even exist yet.

```sql
BEGIN TRANSACTION;
CREATE TABLE "posts" (id int primary key);
INSERT INTO "posts" VALUES(0);
DELETE FROM "sqlite_sequence";
INSERT INTO "sqlite_sequence" VALUES('tags',1);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(1,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
COMMIT;
```

If we now try to run the sql commands pythons iterdump has provided us with in a completely fresh db:
```sql
sqlite> BEGIN TRANSACTION;
CREATE TABLE "posts" (id int primary key);
INSERT INTO "posts" VALUES(0);
DELETE FROM "sqlite_sequence";
INSERT INTO "sqlite_sequence" VALUES('tags',1);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(1,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
COMMIT;
Error: no such table: sqlite_sequence
```

We get an error because the sqlite_sequence table does not exist yet. That's why its necessary for the `sqlite_sequence` related sql commands to be at the very end of the sql script that iterdump and .dump produce.

erlend-aasland replied:
Great job, thanks for investigating!

I think we should audit the whole iterdump command against the sqlite3 cli .dump command, but that is for another PR.

erlend-aasland commented on Lib/sqlite3/test/test_dump.py:
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
+            """id integer primary key autoincrement,"""
+            """tag varchar(256),"""
+            """post int references posts);"""
+        ]
```
Can you wrap each statement on a line for itself. For one-liner string literals, please just use a pair of double-quotes for the string literal.

Also, would it be easier to just define this as a string with newlines, instead of a list of strings? For example:
```python3
        expected_sqls = """
            create table "posts" (it int primary key);
            insert into "posts" values(0);
            create table "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
            insert into "tags" values(NULL, 'test', 0);
            create table "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
        """

        self.cu.executescript(expected_sqls)

        # etc...
```

Ditto for `test_dump_autoincrement_create_new_db`.

itssme replied:
has been changed in newest commit

erlend-aasland commented on Lib/sqlite3/test/test_dump.py:
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
+        for sql_statement in old_db:
+            self.cu.execute(sql_statement)
+
+        cx2 = sqlite.connect(":memory:")
```
Add `from .test_dbapi import managed_connect` to the top of `test_dump.py`, and use it here. See examples in `Lib/sqlite3/test/test_dbapi.py`.

```python3
        with managed_connect(":memory:", in_mem=True) as cx:
            ...
```

You can then drop the explicit closing of `cu2` and `cx2`

itssme replied:
has been changed to `with memory_database() as cx2:` according to the implementation and examples in `test_dbapi.py`

erlend-aasland commented on Lib/sqlite3/test/test_dump.py:
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
+                        ['COMMIT;']
+
+        for i in range(len(expected_sqls)):
+            self.assertEqual(expected_sqls[i], actual_sqls[i])
```
```suggestion
        self.assertEqual(expected_sqls, actual_sqls)
```

itssme replied:
has been changed in newest commit

erlend-aasland commented on Lib/sqlite3/test/test_dump.py:
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
+            self.cu.execute(sql_statement)
+
+        cx2 = sqlite.connect(":memory:")
+        query = "".join(line for line in self.cx.iterdump())
```
```suggestion
        query = "".join(self.cx.iterdump())
```

itssme replied:
has been changed in newest commit

erlend-aasland reviewed (changes requested):
Thanks for updating your PR :) I've left some comments.

itssme committed:
Merge branch 'main' into fix-issue-34828

itssme committed:
bpo-34828: reformatted some tests in test_sqlite3/test.dump.py as well as small formatting changes in sqlite3/dump.py

erlend-aasland committed:
Merge branch 'main' into fix-issue-34828

erlend-aasland commented on Lib/test/test_sqlite3/test_dump.py:
```diff
@@ -49,6 +51,59 @@ def test_table_dump(self):
+
+            self.assertEqual(cu2.execute("SELECT \"seq\""
+                                         "FROM \"sqlite_sequence\""
+                                         "WHERE \"name\" == \"tags\"").fetchall()[0][0], 9)
```
This will not work if SQLite is compiled with, for example, `-DSQLITE_DQS=0`; you need to use single quotes around string literals. Ditto for the other SELECT.

itssme replied:
Thanks for the hint, I was not aware of that. It should be fixed in the newest version :+1:

erlend-aasland reviewed (commented):
Thanks for doing this!

I think the tests can be simplified; the `posts` table is not really needed to test this functionality, so I suggest to drop it, in order to narrow down the tests.

In `test_dump_autoincrement_create_new_db`, one test table should be sufficient, right?

I also got some style nits; do you mind if I just apply them before merge?

itssme committed:
bpo-34828: minified test case for some sqlite3 dump command tests, fixed wrong use of double quotes and single quotes in sqlite3 statements

erlend-aasland committed:
Merge branch 'main' into fix-issue-34828

erlend-aasland committed:
Clean up dump.py

- consolidate construction of sqlite_sequence
- use existing quoting style

erlend-aasland committed:
Cleanup: don't use abbreviated link in NEWS entry

erlend-aasland committed:
Cleanup: tests

- reduce number of escape chars in string literals
- normalise variable names
- in test_dump_autoincrement_create_new_db, simplify setup
- use subTest in order to make it easier to assert more thoroughly

erlend-aasland committed:
Fix  1abc7d05b7a605224611a09eaa74e50ef1cc38e1
